### PR TITLE
HTTP Client API: GET /account/balance at different blocks

### PR DESCRIPTION
**Pull request by velzevur**
[Pivotal story](https://www.pivotaltracker.com/story/show/154061138)
### Description
Get account's balance with an optional block height or hash

#### Get at top
```
$ curl -G 'http://127.0.0.1:3113/v2/account/balance/ak$3du5GfVu6VbDaK2VPH99pcVXaf5rUudNjuDKay44D3MaVCmNhf6oARkdmS2jnmuzkf69BUDTLzdeqNjMo3KgNmDB9Ssxwg'
{"balance":14340}
```

#### Get at height 30
```
$ curl -G 'http://127.0.0.1:3113/v2/account/balance/ak$3du5GfVu6VbDaK2VPH99pcVXaf5rUudNjuDKay44D3MaVCmNhf6oARkdmS2jnmuzkf69BUDTLzdeqNjMo3KgNmDB9Ssxwg' --data-urlencode 'height=30'
{"balance":300}
```

#### Get at height 30 by hash
Hash of block with height 30 is `bh$2Yrk91ysDS8X4FgULJi3z8RKCaVPV1KUX73iBLXkGooXugY3L1`
```
$ curl -G 'http://127.0.0.1:3113/v2/account/balance/ak$3du5GfVu6VbDaK2VPH99pcVXaf5rUudNjuDKay44D3MaVCmNhf6oARkdmS2jnmuzkf69BUDTLzdeqNjMo3KgNmDB9Ssxwg' --data-urlencode 'hash=bh$2Yrk91ysDS8X4FgULJi3z8RKCaVPV1KUX73iBLXkGooXugY3L1'
{"balance":300}
```

### CODE REVIEW

**Comment by velzevur** [comment]
Left to do:

- [ ] Modify protocol repo to the changed API (no longer an external `GET /account/balance` endpoint)
- [ ] Add a detailed description in wiki

**Comment by velzevur** [comment]
[Wiki description](https://github.com/aeternity/epoch/wiki/HTTP-Client-API:-get-account-balance)

File: apps/aehttp/src/aehttp_dispatch_int.erl [hunk 1]
```diff
@@ -203,7 +203,7 @@ handle_request('PostNameUpdateTx', #{'NameUpdateTx' := NameUpdateTxObj}, _Contex
                             nonce     => Nonce,
                             name_hash => DecodedNameHash,
                             name_ttl  => NameTTL,
-                            pointers  => jsx:decode(Pointers),
+                            pointers  => jsx:decode(Pointers,[{labels, atom}]),
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_dispatch_int.erl]: shouldn't it be `{labels, existing_atom}`?

Comment (hanssv) [inline apps/aehttp/src/aehttp_dispatch_int.erl]: Wasn't the atoms supposed to be removed from naming?

File: apps/aehttp/src/aehttp_dispatch_int.erl [hunk 2]
```diff
@@ -386,6 +386,29 @@ handle_request('GetTxsListFromBlockRangeByHash', Req, _Context) ->
             {400, [], #{reason => <<"Invalid hash">>}}
     end;
           
+
+handle_request('GetAccountBalance', Req, _Context) ->
+    case aec_base58c:safe_decode(account_pubkey, maps:get('account_pubkey', Req)) of
+        {ok, AccountPubkey} ->
+            case get_block_hash_optionaly_by_hash_or_height(Req) of
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_dispatch_int.erl]: `_optionally_` ... (spelling)

File: apps/aehttp/src/aehttp_dispatch_int.erl [hunk 3]
```diff
@@ -649,3 +672,57 @@ get_block_range(GetFun, Req) when is_function(GetFun, 0) ->
...
+    GetHashByHeight =
+        fun(Height) ->
+            case aec_conductor:get_header_by_height(Height) of 
+                {error, chain_too_short} ->
+                    {error, not_found};
+                {ok, Header} ->
+                    {ok, _Hash} = aec_headers:hash_header(Header) 
+            end
+        end,
+    case {maps:get('height', Req), maps:get('hash', Req)} of
```
Comment (ThomasArts) [inline apps/aehttp/src/aehttp_dispatch_int.erl]: gives an exception if not there, use maps:get/3

File: apps/aehttp/src/aehttp_dispatch_int.erl [hunk 4]
```diff
@@ -649,3 +672,57 @@ get_block_range(GetFun, Req) when is_function(GetFun, 0) ->
...
+    {value, Trees} = aec_db:find_block_state(Hash),
+    AccountsMPTree = aec_trees:accounts(Trees),
+    case aec_accounts_trees:lookup(AccountPubkey, AccountsMPTree) of
+        none ->
+            {error, account_not_found};
+        {value, Account} ->
+            {ok, aec_accounts:balance(Account)}
+    end.
+
+-spec get_block_hash_optionaly_by_hash_or_height(map()) ->
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_dispatch_int.erl]: `_optionally_` ... (spelling)

File: apps/aehttp/src/aehttp_dispatch_int.erl (hunk 1 above)
Comment (velzevur) [reply apps/aehttp/src/aehttp_dispatch_int.erl]: I think git just backstabbed me. Thanks

Comment (ThomasArts) [reply apps/aehttp/src/aehttp_dispatch_int.erl]: Rebase to master should fix this.

File: apps/aehttp/src/aehttp_dispatch_int.erl (hunk 3 above)
Comment (ThomasArts) [reply apps/aehttp/src/aehttp_dispatch_int.erl]: A test should have caught this... why did it slip through?

File: config/swagger.yaml [hunk 5]
```diff
@@ -552,6 +525,43 @@ paths:
...
+        - application/json
+      parameters:
+        - in: path 
+          name: account_pubkey
+          description: 'Account pubkey to show balance for'
+          required: true
+          type: string
+        - in: query 
+          name: height
+          description: 'Height of the block show balance at'
```
Comment (uwiger) [inline config/swagger.yaml]: "to show ..."

File: apps/aehttp/src/aehttp_dispatch_int.erl (hunk 3 above)
Comment (velzevur) [reply apps/aehttp/src/aehttp_dispatch_int.erl]: Actually this is a swagger thingy. All params that are not required are put in the Req (having a value `undefined`), even if not provided in the request. Using a maps:get/3 would never have the default param being used.

File: config/swagger.yaml [hunk 6]
```diff
@@ -552,6 +525,43 @@ paths:
...
+          required: false
+          type: integer
+        - in: query 
+          name: hash
+          description: 'Hash of the block show balance at'
```
Comment (uwiger) [inline config/swagger.yaml]: "to show ..."

File: apps/aehttp/src/aehttp_dispatch_int.erl (hunk 3 above)
Comment (ThomasArts) [reply apps/aehttp/src/aehttp_dispatch_int.erl]: Ok, that happens under water... good to know

**Comment by velzevur** [comment]
Addressed comments in a seperate commit. I will squash those at merge.

**Comment by uwiger** [comment]
From the Travis run:
```
===> Compiling _build/test/lib/aehttp/test/aehttp_integration_SUITE.erl failed
_build/test/lib/aehttp/test/aehttp_integration_SUITE.erl:1458: function get_balance/0 undefined
_build/test/lib/aehttp/test/aehttp_integration_SUITE.erl:1470: function get_balance/0 undefined
```

**Comment by velzevur** [comment]
CI failed
```
_build/test/lib/aehttp/test/aehttp_integration_SUITE.erl:1458: function get_balance/0 undefined
_build/test/lib/aehttp/test/aehttp_integration_SUITE.erl:1470: function get_balance/0 undefined
```
I've renamed the function to get_balance_at_top/0 but the old name was used in some NS tests. Force pushed

**Comment by velzevur** [comment]
Corner case handeling: if one requests for an orphan block or a block that is part of a side chain - we will still have a valid hash of the block but no corresponding state will be present in the state's DB, thus `aec_db:find_block_state/1` returning `none`. Latest commit addresses that situation
